### ISSUE TITLE
Nested Attributes get not serialized correctly

### ISSUE THREAD

**Issue by kaelumania** [DmitryTsepelev/store_model#60]
Hello,

I was wondering, why during serialization my custom types don't get used to determine the json format:

```ruby
class DateOrTimeType < ActiveRecord::Type::Value
  def type
    :json
  end

  def cast_value(value)
    case value
    when String
      decoded = ActiveSupport::JSON.decode(value) rescue nil
      build_from(decoded)
    when Hash
      build_from(value)
    when Date, Time, DateTime
      value
    end
  end

  def serialize(value)
    case value
    when Date
      ActiveSupport::JSON.encode(date: value)
    when Time, DateTime
      ActiveSupport::JSON.encode(datetime: value)
    else
      super
    end
  end

  def changed_in_place?(raw_old_value, new_value)
    cast_value(raw_old_value) != new_value
  end

  private

  def build_from(hash)
    if hash['date'].present?
      Date.parse(hash['date'])
    else
      Time.parse(hash['datetime'])
    end
  end

end
```

```ruby
class Appointments::Schedule
  include StoreModel::Model

  attribute :from, DateOrTimeType.new
  attribute :to, :date_or_time
```

but during serialization, the default json serialisation of the given type is used.

**Comment by DmitryTsepelev** [comment]
Hi @kaelumania! Could you please provide the example of what format you're trying to get and what you see instead? Gist or failing spec would be really helpful

**Comment by DmitryTsepelev** [comment]
Closing the issue for now

**Comment by rmckayfleming** [comment]
I’m running into a similar issue. As far as I can tell, serialize and deserialize are never called on attributes with custom types. The only method called is cast.

**Comment by DmitryTsepelev** [comment]
Yeah, I re–checked it quickly, and looks like a regular `#save` calls only `cast`/`cast_value`

**Comment by flop** [comment]
I just ran into this problem. Serialize is not called on save. Here is a simplified version of what I'm trying to do : 

```ruby
class Shipment < ActiveRecord::Base
  attribute :recipient, Shipment::Recipient.to_type
end

class Shipment::Recipient
  include StoreModel::Model
  attribute :country, :country
end

require 'countries' # https://github.com/countries/countries
class CountryType < ActiveModel::Type::Value
  def cast(value)
    ISO3166::Country.new(value)
  end

  def serialize(value)
    value.alpha2
  end
end

shipment = Shipment.create!(recipient: {country: 'FR'})
```
This should save {"country": ''FR'} in the recipient attribute in the database and shipment.recipient.country should be a ISO3166::Country. But instead the gem is trying to save a hash of all the data from ISO3166::Country without calling serialize.

The CountryType works correctly when used with a direct string attribute.

**Comment by hallelujah** [comment]
Yes this is because of https://github.com/DmitryTsepelev/store_model/blob/44071d2f5b3367ab19529438dd56fe02f2149a77/lib/store_model/types/one.rb#L49-L50

TL;DR It calls directly `as_json` on the passed hash instead of serializing the values as defined.

**Comment by makikata** [comment]
I worked it around somehow including `ActiveModel::Serializers::JSON` and overriding `read_attribute_for_serialization`

```ruby
class Message < ApplicationRecord
  attribute :payload, MessagePayload.to_type
end

class MessagePayload
  include StoreModel::Model
  # Workaround for: https://github.com/DmitryTsepelev/store_model/issues/60#issuecomment-962668573
  # `as_json` will read values from `read_attribute_for_serialization`
  include ActiveModel::Serializers::JSON

  # 
  attribute :user, ActiveModel::Type::GlobalId.new

  def read_attribute_for_serialization(attribute)
    attribute_types[attribute].serialize(super)
  end


module ActiveModel
  module Type
    # An ActiveModel::Type that serializes and deserializes GlobalID object    
    class GlobalId < Value

      def type
        :global_id
      end

      def serialize(value)
        value&.to_gid&.to_s
      end

      def cast_value(value)
        value.is_a?(::String) ? GlobalID::Locator.locate(value) : value
      end

      def assert_valid_value(value)
        value.nil? ||  value.respond_to?(:to_gid) || value.to_s.start_with?('gid://')
      end
    end
  end
end
```

**Comment by 23tux** [comment]
Is there any news on this? I'm running into a similar problem, when trying to build a type that encrypts data when stored to the DB and encrypts it on the fly when read.

**Comment by DmitryTsepelev** [comment]
@23tux @flop @kaelumania could you please take a look at #60? I created a custom type with the serialization method and it seems to work. What am I missing?

**Comment by 23tux** [comment]
@DmitryTsepelev I just stumbled upon this old issue, as I'm trying to implement some encryption handling into StoreModel (again).

My approach seems to be fine for non-OneOf use cases. I tried to make a ready to use file that can be run with `ruby debug.rb`:

```
# frozen_string_literal: true

ENV["BUNDLE_GEMFILE"] = ""
require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", "7.1.3.3"
  gem "sqlite3", "1.7.3"
  gem "store_model", "3.0.0"
end

require "active_record"
require "active_support/all"
require "store_model"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")

ActiveRecord::Schema.define do
  create_table :dummies, force: true do |t|
    t.json :duck
    t.json :pond
    t.json :water
  end
end

class EncryptedString < ActiveRecord::Type::String
  def deserialize(value)
    return unless value = super

    Rails::Secrets.decrypt(value)
  end

  def serialize(value)
    return unless value = super

    Rails::Secrets.encrypt(value)
  end
end

class Duck
  include StoreModel::Model
  attribute :name, EncryptedString.new
end

class Pond
  include StoreModel::Model
  attribute :model, :string
  attribute :duck, Duck.to_type
end

class Lake
  include StoreModel::Model
  attribute :model, :string
  attribute :duck, Duck.to_type
end

Water = StoreModel.one_of do |json|
  json[:model] == "lake" ? Lake : Pond
end

class Dummy < ActiveRecord::Base
  attribute :duck, Duck.to_type
  attribute :pond, Pond.to_type
  attribute :water, Water.to_type
end

dummy = Dummy.create(
  duck: { name: "Steve" },
  pond: { model: "pond", duck: { name: "John" } },
  water: { model: "lake", duck: { name: "Bob" } }
).reload

puts dummy.duck.inspect
=> #<Duck name: "Steve">

puts dummy.pond.inspect
=> #<Pond model: "pond", duck: #<Duck name: "John">>

puts dummy.water.inspect
=> #<Pond model: "lake", duck: #<Duck name: "ucN9gjlKAIHdcHrdhw==--N8nHKnRBj6NL4X6a--SFwr8nj1MfiPZlqykMgRww==">>
```

As you can see, the `water` attribute, which has a `OneOf` configuration,  did NOT successfully decrypt the data. Can you help me find out why? Why isn't the `#deserialize` method called from my custom type?
